cpp/src/io/json/reader_impl.cu

```diff
@@ -111,15 +72,176 @@ constexpr size_t calculate_max_row_size(int num_columns = 0) noexcept
+  cudf::io::json::gpu::collect_keys_info(data_ptr,
+                                         data.size(),
+                                         opts,
+                                         record_starts.data().get(),
+                                         record_starts.size(),
+                                         key_counter.data(),
+                                         nullptr,
+                                         stream);
+
+  // Allocate columns to store hash values, lengths, offsets
+  auto const num_keys = key_counter.value();
+  std::vector<std::unique_ptr<column>> info_columns;
+  info_columns.emplace_back(make_numeric_column(data_type(type_id::UINT64), num_keys));
+  info_columns.emplace_back(make_numeric_column(data_type(type_id::UINT16), num_keys));
+  info_columns.emplace_back(make_numeric_column(data_type(type_id::UINT32), num_keys));
+  auto info_table = std::make_unique<table>(std::move(info_columns));
+  rmm::device_scalar<mutable_table_device_view> info_table_mdv(
+    *mutable_table_device_view::create(info_table->mutable_view(), stream), stream);
+
+  // Reset the key counter - now used for indexing
+  key_counter.set_value(0, stream);
+  // Fill the allocated columns
+  cudf::io::json::gpu::collect_keys_info(data_ptr,
+                                         data.size(),
+                                         opts,
+                                         record_starts.data().get(),
+                                         record_starts.size(),
+                                         key_counter.data(),
+                                         info_table_mdv.data(),
+                                         stream);
+  return std::move(info_table);
+}
+
+/**
+ * @brief Extract the keys from the JSON file the name offsets/lengths.
+ */
+std::vector<std::string> create_key_strings(uint8_t const *d_data_ptr,
+                                            table_view sorted_info,
+                                            cudaStream_t stream)
+{
+  auto const num_cols = sorted_info.num_rows();
+  std::vector<uint64_t> h_offsets(num_cols);
+  cudaMemcpyAsync(h_offsets.data(),
+                  sorted_info.column(0).data<uint64_t>(),
+                  sizeof(uint64_t) * num_cols,
+                  cudaMemcpyDefault,
+                  stream);
+
+  std::vector<uint16_t> h_lens(num_cols);
+  cudaMemcpyAsync(h_lens.data(),
+                  sorted_info.column(1).data<uint16_t>(),
+                  sizeof(uint16_t) * num_cols,
+                  cudaMemcpyDefault,
+                  stream);
+
+  std::vector<std::string> names(num_cols);
+  std::transform(
+    h_offsets.begin(), h_offsets.end(), h_lens.begin(), names.begin(), [&](auto offset, auto len) {
+      std::vector<char> buffer(len);
+      cudaMemcpyAsync(buffer.data(), d_data_ptr + offset, len, cudaMemcpyDefault, stream);
+      return std::string(reinterpret_cast<const char *>(buffer.data()), len);
+    });
```

Comment:
**Suggestion:** copy the `d_data_ptr ` with offset,len once and convert to `std::string`.
```suggestion
  auto d_offsets = sorted_info.column(0).data<uint64_t>();
  auto d_lengths = sorted_info.column(1).data<uint16_t>();
  auto chars_length =
    thrust::reduce(rmm::exec_policy(stream)->on(stream), d_lengths, d_lengths + num_cols);
  rmm::device_uvector<char> dbuffer(chars_length, stream);
  thrust::for_each_n(rmm::exec_policy(stream)->on(stream),
                     thrust::make_zip_iterator(thrust::make_tuple(d_offsets, d_lengths)),
                     num_cols,
                     [dbuffer_ptr = dbuffer.data(), d_data_ptr](auto offset_len) {
                       auto offset = thrust::get<0>(offset_len);
                       auto len    = thrust::get<1>(offset_len);
                       memcpy(dbuffer_ptr + offset, d_data_ptr + offset, len);
                     });
  std::vector<char> buffer;
  buffer.resize(chars_length);  // To avoid initialization
  CUDA_TRY(cudaMemcpyAsync(buffer.data(), dbuffer.data(), chars_length, cudaMemcpyDefault, stream));
  std::transform(
    h_offsets.begin(), h_offsets.end(), h_lens.begin(), names.begin(), [&](auto offset, auto len) {
      return std::string(reinterpret_cast<const char*>(buffer.data() + offset, len), len);
    });
```

Reply:
I found that this data is also available on host. The new version does not involve any d2h copies.